react/features/base/tracks/actions.js

```diff
@@ -89,9 +95,46 @@ export function createLocalTracksA(options = {}) {
                 /* firePermissionPromptIsShownEvent */ false,
                 store)
             .then(
-                localTracks => dispatch(_updateLocalTracks(localTracks)),
-                reason =>
-                    dispatch(_onCreateLocalTracksRejected(reason, device)));
+                localTracks => {
+                    // Because GUM is called for 1 device (which is actually
+                    // a media type 'audio','video', 'screen' etc.) we should
+                    // not get more than one JitsiTrack.
+                    if (localTracks.length !== 1) {
+                        throw new Error(
+                            'Expected exactly 1 track, but was '
+                                + `given ${localTracks.length} tracks`
+                                + `for device: ${device}.`);
+                    }
+
+                    if (gumProcess.canceled) {
+                        return _disposeTracks(localTracks)
+                                .then(dispatch(_trackCreateCanceled(device)));
+                    }
+
+                    return dispatch(trackAdded(localTracks[0]));
+                },
+                // eslint-disable-next-line prefer-arrow-callback
+                function(reason) {
+                    return gumProcess.canceled
+                        ? dispatch(_trackCreateCanceled(device))
+                        : dispatch(
+                            _onCreateLocalTracksRejected(reason, device));
+                });
+
+            gumProcess.cancel = () => {
+                gumProcess.canceled = true;
+
+                return gumProcess;
+            };
+
+            dispatch({
+                type: TRACK_BEING_CREATED,
+                track: {
```

Comment:
How about adding a `dispose` function here so the API to dispose a track is consistent, regardless of it being a stub track?

Reply:
adding 'dispose' to what ? or do you mean renaming 'cancel' to 'dispose' ? Note that it will not be consistent anyway, because it's not a JitsiTrack and there must not be a 'jitsiTrack' field on that stub.

Reply:
The dispose function would do what the current cancel does, mark it as cancelled. Just an idea.

Reply:
@lyubomir do you have any opinion on that one ? I personally would rather not do that, because the gumProcess is something different than a JitsiTrack and it would be confusing in my opinion (method dispose on a process does not make sense to me). It may slightly simplify the disposeLocalTracks function if we remap both 'gumProcess' and 'jitsiTrack' and use it as an object that has dispose method later.

Reply:
* A "process" is more naturally to me cancelled, interrupted rather than disposed of.
* We already used "thenable with cancel" in lib-jitsi-meet for authentication.

Reply:
@saghul, did you mean having the value of `track` equal to an object (it's already like that) with a method `dispose` defined on it (that's the new part that I think you may have wanted)? And the `track.dispose` would call `gumProcess.cancel` in this case or `jitsiTrack.disposeOrWhateverItIsCalled` in the other case?